motion_control/controllers/profile_feedforward_controller/ProfileFeedforwardController.cpp

```diff
@@ -0,0 +1,231 @@
+    if (recompute_profile) {
+        DEBUG("ProfileFeedforward[%s]: RECOMPUTE requested, pose_error=%.2f\n",
+              keys_.pose_error.data(), static_cast<double>(pose_error));
+        // Use pose error as target distance
+        float target_distance = pose_error;
+
+        // Store direction sign and use absolute distance for profile generation
+        // The sign will be applied to feedforward velocity output
+        direction_sign_ = (target_distance >= 0.0f) ? 1.0f : -1.0f;
+        float absolute_distance = etl::absolute(target_distance);
+
+        // Read current speed (use absolute value, clamped to max_speed)
+        float current_speed = 0.0f;
+        if (auto opt = io.get_as<float>(keys_.current_speed)) {
+            current_speed = etl::absolute(*opt);
+            // Clamp to max_speed to prevent corrupted odometry from creating invalid profiles
+            if (current_speed > parameters_.max_speed()) {
+                LOG_WARNING(
+                    "ProfileFeedforwardController: current_speed %.2f clamped to max %.2f\n",
+                    current_speed, parameters_.max_speed());
+                current_speed = parameters_.max_speed();
+            }
+        }
+
+        // Generate optimal profile with absolute distance
+        uint32_t total_periods = profile_.generate_optimal_profile(
+            current_speed, absolute_distance, parameters_.acceleration(),
+            parameters_.deceleration(), parameters_.max_speed(), parameters_.must_stop_at_end());
+
+        if (total_periods == 0) {
+            LOG_WARNING("ProfileFeedforwardController: No profile generated (distance=%.2f)\n",
+                        absolute_distance);
+            io.set(keys_.feedforward_velocity, 0.0f);
+            io.set(keys_.tracking_error, 0.0f);
+            profile_ready_ = false;
+            return;
+        }
+
+        DEBUG("[%s] Generated profile: %" PRIu32 " periods, distance=%.2f, sign=%.0f\n",
+              keys_.pose_error.data(), total_periods, absolute_distance, direction_sign_);
+
+        // Reset period counter
+        period_ = 0;
+        profile_ready_ = true;
+    }
+
+    // If no profile ready, output zero
+    if (!profile_ready_) {
+        io.set(keys_.feedforward_velocity, 0.0f);
+        io.set(keys_.tracking_error, 0.0f);
+        return;
+    }
+
+    // Check if profile is complete
+    bool profile_complete = (period_ >= profile_.total_periods());
+
+    // When profile is complete, continue with PID-only control using pose_error
+    // Don't invalidate - let the state machine handle the transition
+    // The feedforward velocity is 0 (profile done), but tracking_error = pose_error
+    // allows the PID to finish bringing the robot to target
+    // NOTE: Use pose_error directly WITHOUT direction_sign_ because pose_error
+    // from PoseStraightFilter is already properly signed (negative for backward motion)
+    if (profile_complete) {
```

Comment:
If i understand well, this condition can be removed if the sign is totally handler inside Trapezoidal profile as at the end of the profile the `feedforward_velocity` is 0 and `theoretical_remaining' is 0 too.

So the line 196 can compute the same output : `(actual_remaining - 0) = pose_error`.

But maybe there is a sign handling mix between PoseStraightFilter and ProfileFeedforwardController too.

Reply:
done